src/R2DH.jl

```diff
@@ -1,321 +1,484 @@
-    * `Chist`: an array with the history of number of inner iterations
-    * `NonSmooth`: the nonsmooth term
-    * `status`: the status of the solver either `:first_order`, `:max_iter`, `:max_time` or `:exception`
-    * `fk`: the value of the smooth objective at the final iterate
-    * `hk`: the value of the nonsmooth objective at the final iterate
-    * `sqrt_ξ_νInv`: the square root of the ratio of the nonsmooth term to the regularization parameter
-    * `elapsed_time`: the elapsed time
-"""
 function R2DH(
   f::F,
   ∇f!::G,
   h::H,
   D::DQN,
   options::ROSolverOptions{R},
   x0::AbstractVector{R};
-  Mmonotone::Int = 6,
   selected::AbstractVector{<:Integer} = 1:length(x0),
   kwargs...,
 ) where {F <: Function, G <: Function, H, R <: Real, DQN <: AbstractDiagonalQuasiNewtonOperator}
-  start_time = time()
-  elapsed_time = 0.0
-  ϵ = options.ϵa
-  ϵr = options.ϵr
-  neg_tol = options.neg_tol
-  verbose = options.verbose
-  maxIter = options.maxIter
-  maxTime = options.maxTime
-  σmin = options.σmin
-  σk = options.σk
-  η1 = options.η1
-  η2 = options.η2
-  ν = options.ν
-  γ = options.γ
-  θ = options.θ
-
-  local l_bound, u_bound
-  has_bnds = false
-  for (key, val) in kwargs
-    if key == :l_bound
-      l_bound = val
-      has_bnds = has_bnds || any(l_bound .!= R(-Inf))
-    elseif key == :u_bound
-      u_bound = val
-      has_bnds = has_bnds || any(u_bound .!= R(Inf))
-    end
-  end
+  nlp = FirstOrderModel(f, ∇f!, x0)
+  reg_nlp = RegularizedNLPModel(nlp, h, selected)
+  stats = R2DH(
+    reg_nlp,
+    x = x0,
+    D = D,
+    atol = options.ϵa,
+    rtol = options.ϵr,
+    neg_tol = options.neg_tol,
+    verbose = options.verbose,
+    max_iter = options.maxIter,
+    max_time = options.maxTime,
+    σmin = options.σmin,
+    η1 = options.η1,
+    η2 = options.η2,
+    ν = options.ν,
+    γ = options.γ,
+    θ = options.θ,
+    kwargs...,
+  )
 
-  if verbose == 0
-    ptf = Inf
-  elseif verbose == 1
-    ptf = round(maxIter / 10)
-  elseif verbose == 2
-    ptf = round(maxIter / 100)
-  else
-    ptf = 1
+  return stats.solution, stats.iter, nothing
+end
+
+function R2DH(reg_nlp::AbstractRegularizedNLPModel{T, V}; kwargs...) where {T, V}
+  kwargs_dict = Dict(kwargs...)
+  m_monotone = pop!(kwargs_dict, :m_monotone, 6)
+  D = pop!(kwargs_dict, :D, nothing)
+  solver = R2DHSolver(reg_nlp, m_monotone = m_monotone, D = D)
+  stats = GenericExecutionStats(reg_nlp.model)
+  solve!(solver, reg_nlp, stats; kwargs_dict...)
+  return stats
+end
+
+function SolverCore.solve!(
+  solver::R2DHSolver{T},
+  reg_nlp::AbstractRegularizedNLPModel{T, V},
+  stats::GenericExecutionStats{T, V};
+  callback = (args...) -> nothing,
+  x::V = reg_nlp.model.meta.x0,
+  atol::T = √eps(T),
+  rtol::T = √eps(T),
+  neg_tol::T = eps(T)^(1 / 4),
+  verbose::Int = 0,
+  max_iter::Int = 10000,
+  max_time::Float64 = 30.0,
+  max_eval::Int = -1,
+  σmin::T = eps(T),
+  η1::T = √√eps(T),
```

Comment:
So according to the last comment, you ought to initialize $$\sigma_k$$ here as well.